File: app/src/main/java/com/policyboss/policybosspro/location/LocationService.java
package com.policyboss.policybosspro.location;

import android.Manifest;
import android.app.Activity;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Build;


import androidx.core.app.ActivityCompat;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.location.LocationSettingsRequest;
import com.google.android.gms.location.LocationSettingsResponse;
import com.google.android.gms.location.SettingsClient;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;

/**
 * Created by Rajeev Ranjan on 08/01/2020.
 */
public class LocationService {

    private Activity mContext;
    ILocation locationListner;

    private LocationRequest locationRequest;

    //  private LocationCallback locationCallback ;

    public static final int MY_PERMISSION_REQUEST_FINE_LOCATION = 101;
    public static final int REQUEST_CHECK_SETTINGS = 0x1;

    FusedLocationProviderClient fusedLocationProviderClient;

    LocationCallback locationCallback = new LocationCallback() {
        @Override
        public void onLocationResult(LocationResult locationResult) {
            super.onLocationResult(locationResult);
            locationListner.getLocation(locationResult.getLastLocation());

        }
    };

    public LocationService(Activity context) {
        mContext = context;
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext);
    }

    public void setLocationListener(ILocation locationListner) {
        this.locationListner = locationListner;
    }

    public void createLocationRequest() {
        locationRequest = LocationRequest.create();
        locationRequest.setInterval(10000);
        locationRequest.setFastestInterval(5000);
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        LocationSettingsRequest.Builder builder = new LocationSettingsRequest.Builder()
                .addLocationRequest(locationRequest);

        SettingsClient client = LocationServices.getSettingsClient(mContext);
        Task<LocationSettingsResponse> task = client.checkLocationSettings(builder.build());

        task.addOnSuccessListener(mContext, new OnSuccessListener<LocationSettingsResponse>() {
            @Override
            public void onSuccess(LocationSettingsResponse locationSettingsResponse) {
               // Toast.makeText(mContext, "addOnSuccessListener", Toast.LENGTH_SHORT).show();

                startLocationUpdates();
            }
        });

        task.addOnFailureListener(mContext, new OnFailureListener() {
            @Override
            public void onFailure( Exception e) {
               // Toast.makeText(mContext, "addOnFailureListener", Toast.LENGTH_SHORT).show();
                if (e instanceof ResolvableApiException) {
                    // Location settings are not satisfied, but this can be fixed
                    // by showing the user a dialog.
                    try {
                        // Show the dialog by calling startResolutionForResult(),
                        // and check the result in onActivityResult().
                        ResolvableApiException resolvable = (ResolvableApiException) e;
                        resolvable.startResolutionForResult(mContext,
                                REQUEST_CHECK_SETTINGS);
                    } catch (IntentSender.SendIntentException sendEx) {
                        // Ignore the error.
                    }
                }
            }
        });
    }

    public void startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(mContext, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {


            fusedLocationProviderClient.requestLocationUpdates(locationRequest, locationCallback, null);

        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                mContext.requestPermissions(new String[]{Manifest.permission.ACCESS_FINE_LOCATION}, MY_PERMISSION_REQUEST_FINE_LOCATION);
            }
        }

    }

    public void stopLocationUpdates() {
        if (locationCallback != null) {
            fusedLocationProviderClient.removeLocationUpdates(locationCallback);
        }

    }

    public interface ILocation {

        void getLocation(Location location);
    }


}

File: app/src/main/java/com/policyboss/policybosspro/utility/LocationService.kt
package com.policyboss.policybosspro.utility

import android.Manifest
import android.app.Activity
import android.content.IntentSender.SendIntentException
import android.content.pm.PackageManager
import android.location.Location
import android.os.Build
import android.widget.Toast
import androidx.core.app.ActivityCompat

import com.google.android.gms.common.api.ApiException
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.location.*


class LocationService(context: Activity, locationListner: ILocation?) {

    var mContext: Activity
    var locationListner: ILocation? = null

    lateinit var locationRequest: LocationRequest
    //var  locationRequest: LocationRequest
    //  private LocationCallback locationCallback ;

    //  private LocationCallback locationCallback ;
    //  private LocationCallback locationCallback ;
    val MY_PERMISSION_REQUEST_FINE_LOCATION: Int = 101
    val REQUEST_CHECK_SETTINGS = 0x1

    var fusedLocationProviderClient: FusedLocationProviderClient? = null

    init {

        mContext = context
        fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(mContext!!)

        this.locationListner = locationListner

//        locationRequest = LocationRequest.create()
//        locationRequest.setInterval(10000)
//        locationRequest.setFastestInterval(5000)
//        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)

    }


    var locationCallback: LocationCallback = object : LocationCallback() {
        override fun onLocationResult(locationResult: LocationResult) {
            super.onLocationResult(locationResult)

            if(locationListner != null){

                locationListner?.getLocation(locationResult.lastLocation) // calling Interface method
            }

        }
    }


//    fun setLocationListener(locationListner: ILocation?) {
//        this.locationListner = locationListner
//    }


    fun createLocationRequest() {

        locationRequest = LocationRequest.create()
        locationRequest.setInterval(10000)
        locationRequest.setFastestInterval(5000)
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)

        val builder = LocationSettingsRequest.Builder()
            .addLocationRequest(locationRequest)


        builder.setAlwaysShow(true)

        val client = LocationServices.getSettingsClient(mContext)
        val task = client.checkLocationSettings(builder.build())
        task.addOnSuccessListener(mContext!!) {
           // Toast.makeText(mContext, "addOnSuccessListener", Toast.LENGTH_SHORT).show()
            startLocationUpdates()
        }

        task.addOnFailureListener(mContext!!) { e ->
           // Toast.makeText(mContext, "No Location Found, Please try Again!!", Toast.LENGTH_SHORT).show()
            if (e is ResolvableApiException) {
                // Location settings are not satisfied, but this can be fixed
                // by showing the user a dialog.
                try {
                    // Show the dialog by calling startResolutionForResult(),
                    // and check the result in onActivityResult().
                    e.startResolutionForResult(
                        mContext!!,
                        REQUEST_CHECK_SETTINGS
                    )
                } catch (sendEx: SendIntentException) {
                    // Ignore the error.
                }
            }
        }
    }

    //Not in used
    private fun turnOnGPS() {
        val builder = LocationSettingsRequest.Builder()
            .addLocationRequest(locationRequest)
        builder.setAlwaysShow(true)
        val result =
            LocationServices.getSettingsClient(mContext)
                .checkLocationSettings(builder.build())
        result.addOnCompleteListener { task ->
            try {
                val response = task.getResult(
                    ApiException::class.java
                )
                Toast.makeText(mContext, "GPS is already tured on", Toast.LENGTH_SHORT)
                    .show()
            } catch (e: ApiException) {
                when (e.statusCode) {
                    LocationSettingsStatusCodes.RESOLUTION_REQUIRED -> try {
                        val resolvableApiException = e as ResolvableApiException
                        resolvableApiException.startResolutionForResult(mContext, 2)
                    } catch (ex: SendIntentException) {
                        ex.printStackTrace()
                    }
                    LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> {}
                }
            }
        }
    }


    fun startLocationUpdates() {
        if (ActivityCompat.checkSelfPermission(
                mContext!!,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            fusedLocationProviderClient!!.requestLocationUpdates(
                locationRequest,
                locationCallback,
                null
            )
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                mContext!!.requestPermissions(
                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                    MY_PERMISSION_REQUEST_FINE_LOCATION
                )
            }
        }
    }

    fun stopLocationUpdates() {
        if (locationCallback != null) {
            fusedLocationProviderClient!!.removeLocationUpdates(locationCallback)
        }
    }




    interface ILocation {
        fun getLocation(location: Location?)
    }

}